### PR TITLE
throw error on startup if one occurs

### PR DESCRIPTION
**Pull request by jmcc0nn3ll**
@meltsufin This _should_ fail the startup of the server if the WebAppContext has issues on startup.

### CODE REVIEW

**Comment by meltsufin** [comment]
Thanks, Jesse! I'll test it out next week.

**Comment by gregw** [comment]
@jmcc0nn3ll Your patch does indeed allow an init exception to be thrown externally to the context.  However because this context is being started by the deployer an not from a direct embedded usage, the exception does not cause the server to exit.       I'll have a look at how we can configure the deployer to achieve that.... stand by....

**Comment by gregw** [comment]
I've updated this PR with a DeploymentCheck class that is attached to the server life cycle.  Once the server is started, it checks for a started Connector and Context, if either is not found then an ISE is thrown which does stop the server...  it also has revealed the first bug in 9.4.0 that we have a NPE when reporting that ISE, which is a cut and paste problem that will be fixed in 9.4.1 - for now it just produced an unnecessary extra exception as the server exits.

File: jetty9-base/src/main/java/com/google/cloud/runtimes/jetty9/DeploymentCheck.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.eclipse.jetty.util.component.LifeCycle;
+
+public class DeploymentCheck extends AbstractLifeCycle.AbstractLifeCycleListener {
+  @Override
+  public void lifeCycleStarted(LifeCycle bean) {
+    if (bean instanceof Server) {
+      Server server = (Server)bean;
+      Connector[] connectors = server.getConnectors();
+      if (connectors.length == 0 || !connectors[0].isStarted()) {
```
Comment (meltsufin) [inline jetty9-base/src/main/java/com/google/cloud/runtimes/jetty9/DeploymentCheck.java]: Will there always be just one connector, or should we loop through them all?

File: jetty9-base/src/main/jetty-base/etc/gae-web.xml [hunk 2]
```diff
@@ -2,6 +2,9 @@
 <!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_9_3.dtd">
 <Configure class="org.eclipse.jetty.webapp.WebAppContext">
 
+  <!-- Throw Exception on Start -->
```
Comment (meltsufin) [inline jetty9-base/src/main/jetty-base/etc/gae-web.xml]: This comment is a little confusing. Can we put something more descriptive like "Propagate application startup exceptions to server", or whatever makes sense.

**Comment by gregw** [comment]
Several reasons why deployer does not propagate errors by default:  some users simply do not want it;  for hot deploy there is nowhere to propagate to.

However, we will look at making start errors able to be propagated.

File: jetty9-base/src/main/java/com/google/cloud/runtimes/jetty9/DeploymentCheck.java (hunk 1 above)
Comment (gregw) [inline jetty9-base/src/main/java/com/google/cloud/runtimes/jetty9/DeploymentCheck.java]: There can be more connectors.   I was just testing minimal conditions of at least 1 working connector, which I guess may not be the first.    So I can test all, but do we check for all started, or just 1 of them started?

Comment (meltsufin) [inline jetty9-base/src/main/java/com/google/cloud/runtimes/jetty9/DeploymentCheck.java]: I guess, all of them? What are the other connectors that we will normally have?

Comment (gregw) [inline jetty9-base/src/main/java/com/google/cloud/runtimes/jetty9/DeploymentCheck.java]: Normally we only have 1.  But a user may extend to have other connectors for private or other purposes.  I'll check that all connectors are started.

**Comment by gregw** [comment]
@meltsufin I can do for async-support.

Note that the CLA checker is confused again and It's saying I can't yet merge?

**Comment by gregw** [comment]
@meltsufin do you want this for https://github.com/GoogleCloudPlatform/appengine-java-vm-runtime/tree/async-support or https://github.com/GoogleCloudPlatform/jetty-gae-compat-runtime?

I could just copy the class into async-cupport, or I could consume a maven artefact produced by jetty-runtime?

**Comment by meltsufin** [comment]
@gregw Can you just copy the class and config to https://github.com/GoogleCloudPlatform/appengine-java-vm-runtime/tree/async-support?

I'll try to fix the CLA checks.

**Comment by meltsufin** [comment]
Sorry @gregw had to squash your commit into Jesse's to make this work. :)

**Pull request merged**